feat: add tick generation options to view options

TCL735 opened the pull request:
Closes #238 

Adds options to View Options to allow user to create ticks on each axis for Line, Band, Heatmap, and Scatter plots. Mosaic plot gets x-axis only.

The options are total ticks, tick start, and tick step. They may be used in any combination. The options look like this:
<img width="1680" alt="Screen Shot 2020-11-08 at 6 12 50 PM" src="https://user-images.githubusercontent.com/10736577/98492931-2284e580-21ee-11eb-9dc3-a05403423987.png">

hoorayimhelping commented on src/shared/components/axisTicks/TimeTickInput.tsx:
```diff
@@ -0,0 +1,236 @@
+            visible={isDatePickerOpen}
+            showEvent={PopoverInteraction.None}
+            hideEvent={PopoverInteraction.None}
+            distanceFromTrigger={8}
```
magic number! what's 8 mean?

TCL735 replied:
According to clockface, this is the pixel distance between the trigger (our button) and the popover's dialog box.

All of the `Popover`s in ui have `8` as the value. I just copied it. Perhaps we can clean that up all in one pass rather than piecemeal. Can we leave this for another PR?

hoorayimhelping replied:
>Can we leave this for another PR?

sure, sounds like a good idea

hoorayimhelping commented on src/shared/components/axisTicks/TimeTickInput.tsx:
```diff
@@ -0,0 +1,236 @@
+  const allowOnClickOutside = () => setIsOnClickOutsideHandlerActive(true)
+  const suppressOnClickOutside = () => setIsOnClickOutsideHandlerActive(false)
+
+  const style: CSSProperties = isDatePickerOpen
```
i think conventionally it's called `styles`

TCL735 replied:
`styles` it is. Done.

hoorayimhelping commented on src/shared/components/axisTicks/TimeTicksOptions.tsx:
```diff
@@ -0,0 +1,55 @@
+    setOptions,
+  } = props
+  return (
+    <>
```
```suggestion
```

this isn't needed right? `Grid.Row` contains all the `ValueTickInputs`

TCL735 replied:
Correct, not needed. I had changed back and forth between a few designs. Forgot to remove the fragment tag. Done now.

hoorayimhelping commented on src/shared/components/axisTicks/TimeTicksOptions.tsx:
```diff
@@ -0,0 +1,55 @@
+          setOptions={setOptions}
+        />
+      </Grid.Row>
+    </>
```
```suggestion
```

hoorayimhelping commented on src/shared/components/axisTicks/ValueTickInput.tsx:
```diff
@@ -0,0 +1,92 @@
+      ])
+    } else {
+      setOptions('GenerateAxisTicks', filteredTickOptions)
+      if (tickOptionInput !== '') {
```
same comment here about the structure of this `if` statement

hoorayimhelping commented on src/shared/components/axisTicks/ValueTicksOptions.tsx:
```diff
@@ -0,0 +1,53 @@
+    setOptions,
+  } = props
+  return (
+    <>
```
```suggestion
```
same thing here with `Grid.Row` being a container, right?

hoorayimhelping commented on src/shared/utils/convertUserInput.test.ts:
```diff
@@ -67,3 +70,67 @@ describe('convertUserInputToNumOrNaN', () => {
+    let result = convertUserInputValueToNumOrNaN()
+    expect(typeof result === 'number')
+    expect(result !== result)
+
```
i think you should add an expect for `isNan` like 
```ts
expect(Number.isNan(result)).toBeTruthy()
```
to these tests

(i also prefer `toBeTruthy` because it's more explicit)

TCL735 replied:
oh yeah good call. Done.

TCL735 replied:
And... actually, `expect(result !== result)` doesn't assert anything. It always passes even if the expression inside is false 😅 

So this change was definitely needed, just to get the tests to run correctly.

hoorayimhelping commented on src/shared/utils/convertUserInput.ts:
```diff
@@ -1,5 +1,8 @@
 import React from 'react'
 
+export const convertUserInputValueToNumOrNaN = (value?: any): number =>
+  value === '' ? NaN : Number(value)
```
Why does `''` get converted to `NaN` and not `Number('') // 0`?

TCL735 replied:
This is specifically used when we have user input. An empty string means the user has deleted all of the input field. We should interpret that case as "not a number" rather than 0.

hoorayimhelping commented on src/shared/utils/useAxisTicksGenerator.ts:
```diff
@@ -0,0 +1,31 @@
+  yTickStep,
+}) =>
+  useMemo(() => {
+    if (!isFlagEnabled('axisTicksGenerator')) {
```
we have to be very careful about how this is deployed if this is done through config cat. the way `isFlagEnabled` is written, you can't be sure that the results will be correct, depending on when the function is called in request's lifecycle

hoorayimhelping reviewed (changes requested):
Great work! have a few suggestions / questions

TCL735 commented on src/shared/components/axisTicks/AxisTicksGenerator.tsx:
```diff
@@ -0,0 +1,203 @@
+import {ButtonShape} from '@influxdata/clockface'
+
+const getGenerateAxisTicksOptionsNames = (axisName: string) => {
+  const result = []
+  result.push(`generate${axisName.toUpperCase()}AxisTicks`)
+  result.push(`${axisName.toLowerCase()}TotalTicks`)
+  result.push(`${axisName.toLowerCase()}TickStart`)
+  result.push(`${axisName.toLowerCase()}TickStep`)
+  return result
```
Oops, yeah no reason to do this here. I think this was the remnants of console.log-ing the return value and not using the shorter syntax after removing it.

I will update it to the shorter syntax.

TCL735 commented on src/shared/components/axisTicks/AxisTicksGenerator.tsx:
```diff
@@ -0,0 +1,203 @@
+
+  if (!isFlagEnabled('axisTicksGenerator')) {
+    return null
+  }
```
It is about as high as it can be. It needs to come after all hooks, as early return is not allowed with hooks.

TCL735 replied:
i can move it above `isTimeColumn`

hoorayimhelping replied:
>It needs to come after all hooks, as early return is not allowed with hooks.

🤦 thanks, didn't know this

TCL735 commented on src/shared/components/axisTicks/AxisTicksGenerator.tsx:
```diff
@@ -0,0 +1,203 @@
+    tickStep,
+  } = props
+
+  const [showInputs, setShowInputs] = useState(
```
Sounds good. Done.

TCL735 commented on src/shared/components/axisTicks/AxisTicksGenerator.tsx:
```diff
@@ -0,0 +1,203 @@
+  const {
+    axisName,
+    columnType,
+    generateAxisTicks,
```
yeah, it does sound like a function. However, "generated" is not correct either. Because it really is just a list of options to use for generating the ticks, not the ticks themselves.

TCL735 replied:
I wanted to match the name in the api, so that's how it ended up as "generateAxisTicks".

The bigger issue is that it might be misnamed in the api as well, if "generateAxisTicks" sounds too much like a function.

TCL735 commented on src/shared/components/axisTicks/AxisTicksGenerator.tsx:
```diff
@@ -0,0 +1,203 @@
+    generateAxisTicks: Array.isArray(generateAxisTicks)
+      ? generateAxisTicks
+      : [],
+    totalTicks: hasTotalTicks ? totalTicks : '',
```
This will be feed into a user input field. An empty string represents "no user input".

Even if we don't want to use an empty string here, there is a subtle distinction between "lack of value" versus 0. "lack of value" means, to Giraffe, please calculate the ticks as you see fit. 0 means exactly that -- 0. For example, 0 for total ticks would mean don't label the axis with any ticks. Or, for tick start, 0 would mean start at the value 0 on the axis. (Tick step can also be 0 -- Giraffe handles that gracefully.)

And, so, we use an empty string here to denote "lack of value" which is not the same as 0.

TCL735 commented on src/shared/components/axisTicks/TimeTickInput.tsx:
```diff
@@ -0,0 +1,236 @@
+    tickOptions,
+    initialTickOptionValue,
+    label,
+    placeholder = 'RFC3339',
```
Ah yes, good call. Done.

TCL735 commented on src/shared/components/axisTicks/TimeTickInput.tsx:
```diff
@@ -0,0 +1,236 @@
+
+  const getDatePickerDateTime = () => {
+    const date = new Date(tickOptionInput)
+    if (date.valueOf() === date.valueOf()) {
```
if `date` is an invalid date, then calling .valueOf() will return `NaN` and `NaN` is never equal to itself.

TCL735 replied:
I will update this to use `Number.isNaN()` to be more explicit

TCL735 commented on src/shared/components/axisTicks/TimeTickInput.tsx:
```diff
@@ -0,0 +1,236 @@
+
+  const handleBlur = () => updateTickOption()
+
+  const handleKeyPress = (event: React.KeyboardEvent<HTMLInputElement>) => {
```
Maybe.

Here, we need it because we want to wait until the user types the whole input before attempting to generate the ticks. Without this, as soon as the user types `1` in Tick Step, it would attempt to generate ticks on a step of 1 millisecond, for example. This helped catch a performance issue in giraffe. But we also do not want eager input fields.

So, a long way of saying: yes, in certain cases, we want this.

TCL735 commented on src/shared/components/axisTicks/TimeTickInput.tsx:
```diff
@@ -0,0 +1,236 @@
+      setOptions('GenerateAxisTicks', filteredTickOptions)
+      if (tickOptionInput !== '') {
+        setTickOptionInputStatus(ComponentStatus.Error)
+      }
```
Yes but with a small change. I will need to call `setOptions` with the computedTickOptions no matter what, so that call goes last.

And i think checking convertedValue might be confusing. I changed it to a check of `Number.isNaN` because that is basically what i want to know here.

Also, I changed it to an if - else if structure. i'd rather not check `NaN`ness twice. Just once and the `else if` can check the other condition.

TCL735 commented:
This also requires changes to make it work properly in the Dashboard page.

TCL735 commented:
@hoorayimhelping in addition to addressing the review comments, i also added another commit to fix dashboard cells. Now, they correctly display the user's selected values from the `<AxisTicksGenerator>`
